iwyu.cc

```diff
@@ -2753,10 +2699,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (IsPointerOrReferenceAsWritten(return_type))
       return;
 
-    set<const Type*> merged_blocked = GetProvidedTypesForFnReturn(callee);
-    merged_blocked.insert(blocked_types_.begin(), blocked_types_.end());
-    ValueSaver<set<const Type*>> s(&blocked_types_, merged_blocked);
-    ReportTypeUse(CurrentLoc(), return_type);
+    ReportWithAdditionalBlockedTypes(return_type,
+                                     GetProvidedTypesForFnReturn(callee));
+  }
+
+  void HandleAutocastOnCallSite(clang::CXXConstructExpr* expr) {
+    // 'Autocast' -- calling a one-arg, non-explicit constructor
+    // -- is a special case when it's done for a function call.
+    // iwyu requires the function-writer to provide the #include
+    // for the casted-to type, just so we don't have to require it
+    // here.  *However*, the function-author can override this
+    // iwyu requirement, in which case we're responsible for the
+    // casted-to type.  See IwyuBaseASTVisitor::CanBeProvidedTypeComponent.
+    if (!IsAutocastExpr(current_ast_node()))
+      return;
+    const Type* type = Desugar(expr->getType().getTypePtr());
```

Comment:
Can you add a little comment here as to why this `Desugar` is necessary? It feels counter-intuitive to me, but I know I have absolutely zero intuition for sugar :)

Reply:
I think it is unnecessary because nearly all control flow paths in `ReportTypeUse` desugar the given type as well. The line just moved from [here](https://github.com/include-what-you-use/include-what-you-use/pull/1313/files/d847df1e54a0299bc0d9d39ef9048925a5291e12#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9L2221).